File: app/src/main/java/com/tydeya/familycircle/presentation/ui/livepart/main/details/MainLiveFragment.java
package com.tydeya.familycircle.presentation.ui.livepart.main.details;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.tydeya.familycircle.App;
import com.tydeya.familycircle.R;
import com.tydeya.familycircle.domain.cooperationorganizer.interactor.abstraction.CooperationInteractorCallback;
import com.tydeya.familycircle.domain.cooperationorganizer.interactor.details.CooperationInteractor;
import com.tydeya.familycircle.domain.oldfamilyinteractor.abstraction.FamilyInteractorCallback;
import com.tydeya.familycircle.domain.oldfamilyinteractor.details.FamilyInteractor;
import com.tydeya.familycircle.presentation.ui.livepart.main.details.storiesrecyclerview.OnClickMemberStoryListener;
import com.tydeya.familycircle.presentation.ui.livepart.main.details.cooperationrecyclerview.CooperationRecyclerViewAdapter;
import com.tydeya.familycircle.presentation.ui.livepart.main.details.storiesrecyclerview.FamilyMembersStoriesRecyclerViewAdapter;

import java.util.ArrayList;

import javax.inject.Inject;

import static com.tydeya.familycircle.data.constants.NavigateConsts.BUNDLE_FULL_PHONE_NUMBER;

public class MainLiveFragment extends Fragment implements OnClickMemberStoryListener,
        FamilyInteractorCallback, CooperationInteractorCallback {

    private NavController navController;

    private RecyclerView familyStoriesRecyclerView;
    private FamilyMembersStoriesRecyclerViewAdapter recyclerViewAdapter;

    private RecyclerView cooperationRecyclerView;
    private CooperationRecyclerViewAdapter cooperationRecyclerViewAdapter;

    @Inject
    FamilyInteractor familyInteractor;

    @Inject
    CooperationInteractor cooperationInteractor;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View root = inflater.inflate(R.layout.fragment_main_live_page, container, false);
        navController = NavHostFragment.findNavController(this);

        familyStoriesRecyclerView = root.findViewById(R.id.main_live_page_family_recycler_view);
        cooperationRecyclerView = root.findViewById(R.id.main_live_page_family_live_tape_recycler_view);

        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        App.getComponent().injectFragment(this);
        setFamilyStoriesRecyclerView();
        setCooperationRecyclerView();

    }

    private void setFamilyStoriesRecyclerView() {
        recyclerViewAdapter = new FamilyMembersStoriesRecyclerViewAdapter(getContext(),
                familyInteractor.getActualFamily().getFamilyMembers(), this);

        familyStoriesRecyclerView.setAdapter(recyclerViewAdapter);
        familyStoriesRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(),
                RecyclerView.HORIZONTAL, false));
    }

    private void setCooperationRecyclerView() {
        cooperationRecyclerViewAdapter =
                new CooperationRecyclerViewAdapter(getContext(), new ArrayList<>());

        cooperationRecyclerView.setAdapter(cooperationRecyclerViewAdapter);

        cooperationRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(),
                LinearLayoutManager.VERTICAL, false));
    }

    @Override
    public void onClickMemberStory(int position) {
        Bundle bundle = new Bundle();
        bundle.putString(BUNDLE_FULL_PHONE_NUMBER, familyInteractor.getActualFamily()
                .getFamilyMembers().get(position).getFullPhoneNumber());
        navController.navigate(R.id.familyMemberViewFragment, bundle);
    }

    @Override
    public void memberDataUpdated() {
        recyclerViewAdapter.refreshData(familyInteractor.getActualFamily().getFamilyMembers());
    }

    @Override
    public void cooperationDataFromServerUpdated() {
        cooperationRecyclerViewAdapter.refreshData(cooperationInteractor.getCooperationData());
    }

    @Override
    public void onPause() {
        super.onPause();
        familyInteractor.unsubscribe(this);
        cooperationInteractor.unsubscribe(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        familyInteractor.subscribe(this);
        cooperationInteractor.subscribe(this);
    }
}

File: app/src/main/java/com/tydeya/familycircle/presentation/ui/livepart/main/details/MainLiveFragment.kt
package com.tydeya.familycircle.presentation.ui.livepart.main.details

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat.getColor
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.tydeya.familycircle.R
import com.tydeya.familycircle.data.constants.NavigateConsts.BUNDLE_FULL_PHONE_NUMBER
import com.tydeya.familycircle.databinding.FragmentMainLivePageBinding
import com.tydeya.familycircle.presentation.ui.livepart.main.details.familymembersrecyclerview.FamilyMembersRecyclerViewAdapter
import com.tydeya.familycircle.presentation.ui.livepart.main.details.familymembersrecyclerview.OnClickMemberStoryListener
import com.tydeya.familycircle.presentation.ui.livepart.memberpersonpage.twitterrecycler.TwitterRecyclerViewAdapter
import com.tydeya.familycircle.presentation.viewmodel.familyviewmodel.FamilyViewModel
import com.tydeya.familycircle.presentation.viewmodel.familyviewmodel.FamilyViewModelFactory
import com.tydeya.familycircle.utils.Resource
import com.tydeya.familycircle.utils.extensions.currentFamilyId

class MainLiveFragment : Fragment(), OnClickMemberStoryListener {

    private var _binding: FragmentMainLivePageBinding? = null
    private val binding get() = _binding!!

    private lateinit var familyRecyclerViewAdapter: FamilyMembersRecyclerViewAdapter

    private lateinit var familyViewModel: FamilyViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentMainLivePageBinding.inflate(inflater, container, false)

        familyViewModel = ViewModelProviders
                .of(requireActivity(), FamilyViewModelFactory(requireActivity().currentFamilyId))
                .get(FamilyViewModel::class.java)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initFamilyStoriesRecyclerView()
        initToolbar()
        initTwitterRecycler()
        initPieChart()
    }

    private fun initFamilyStoriesRecyclerView() {
        familyRecyclerViewAdapter = FamilyMembersRecyclerViewAdapter(
                ArrayList(), this
        )
        binding.mainLivePageFamilyRecyclerview.adapter = familyRecyclerViewAdapter
        binding.mainLivePageFamilyRecyclerview.layoutManager = LinearLayoutManager(
                requireContext(), RecyclerView.HORIZONTAL, false)

        familyViewModel.familyMembers.observe(viewLifecycleOwner, androidx.lifecycle.Observer {
            when (it) {
                is Resource.Success -> {
                    familyRecyclerViewAdapter.refreshData(it.data)
                }
                is Resource.Loading -> {
                }
                is Resource.Failure -> {

                }
            }
        })
    }

    private fun initTwitterRecycler() {
        val twitterAdapter = TwitterRecyclerViewAdapter()
        binding.twitterRecycler.adapter = twitterAdapter
        binding.twitterRecycler.layoutManager = LinearLayoutManager(
                requireContext(), LinearLayoutManager.VERTICAL, false
        )
        familyViewModel.tweets.observe(viewLifecycleOwner, Observer {
            when (it) {
                is Resource.Success -> twitterAdapter.refreshTweets(it.data)
                is Resource.Loading -> twitterAdapter.refreshTweets(ArrayList())
            }
        })

        familyViewModel.familyMembers.observe(viewLifecycleOwner, Observer {
            when (it) {
                is Resource.Success -> twitterAdapter.refreshFamilyMembers(it.data)
                is Resource.Loading -> twitterAdapter.refreshFamilyMembers(ArrayList())
            }
        })
    }

    private fun initToolbar() {
        familyViewModel.familyData.observe(viewLifecycleOwner, androidx.lifecycle.Observer {
            when (it) {
                is Resource.Success -> {
                    binding.toolbar.title = it.data.title
                }

                is Resource.Loading -> {
                    binding.toolbar.title = getString(R.string.loading)
                }
            }
        })
    }

    private fun initPieChart() {

        val kitchenData = ArrayList<PieEntry>()
        kitchenData.add(PieEntry(16f, "Fresh food"))
        kitchenData.add(PieEntry(5f, "Ordinary food"))
        kitchenData.add(PieEntry(3f, "Spoiled food"))
        kitchenData.add(PieEntry(10f, "Unknown"))

        val dataSet = PieDataSet(kitchenData, "")

        val data = PieData(dataSet)
        data.setValueTextColor(getColor(requireContext(), R.color.colorWhite))
        data.setValueTextSize(16f)
        binding.kitchenPieChart.data = data
        dataSet.colors = listOf(
                getColor(requireContext(), R.color.colorFreshFood),
                getColor(requireContext(), R.color.colorOrdinaryFood),
                getColor(requireContext(), R.color.colorSpoiledFood),
                getColor(requireContext(), R.color.colorGray))
        binding.kitchenPieChart.animateXY(500, 500)

        binding.kitchenPieChart.legend.isEnabled = true
        binding.kitchenPieChart.centerText = "Shelf life products"
        binding.kitchenPieChart.setDrawEntryLabels(false)
        binding.kitchenPieChart.description.isEnabled = false
    }


    override fun onClickFamilyMember(phoneNumber: String) {
        NavHostFragment.findNavController(this).navigate(
                R.id.familyMemberViewFragment,
                bundleOf(BUNDLE_FULL_PHONE_NUMBER to phoneNumber)
        )
    }

}